src/oc_erchef/apps/chef_objects/src/chef_cookbook_version.erl

```diff
@@ -589,6 +672,79 @@ list_query(_ObjectRec) ->
 bulk_get_query(_ObjectRec) ->
     error(not_implemented).
 
+%% for v2 metadata, the "segment" is encoded in to the name of file
+%% so for data that is v0 on disk we have to prepend the segment to the name
+add_segment_to_filename(Segment, File) ->
+    Fn = ej:get({<<"name">>}, File),
+    Fn1 = iolist_to_binary([Segment, "/", Fn]),
+    ej:set({<<"name">>}, File, Fn1).
+
+remove_segment_from_filename(File) ->
+    [Segment | Name ] = get_segment_from_record(File),
+    case Name of
+        % if Name is nil, then we have a root file (like metadata.rb) - so we'll set the name to the segment, and the segment to <<"root_files">>
+        [] ->
+            { <<"root_files">>, ej:set({<<"name">>}, File, Segment)};
+        _ ->
+            Record = ej:set({<<"name">>}, File, lists:last(Name)),
+            { Segment, Record }
+    end.
+
+populate_all_files(Segment, Data, Metadata) ->
+    ExpandedFiles = case Segment of
+                        <<"root_files">> -> Data;
+                        _ ->
+                            [ add_segment_to_filename(Segment, File) || File <- Data ]
+                    end,
+    ej:set({<<"all_files">>}, Metadata, ExpandedFiles).
+
+%% A v2 cookbook version contains only the "all_files" key, which is a list of all the file parts
+%% { "all_files": [ { "name": "recipes/default.rb", "path": "recipes/default.rb", … } ] }
+%% We have to transform that into segments (?COOKBOOK_SEGMENTS) containing a list of file parts for that segment
+%% { "recipes": [ { "name": "default.rb", path: "recipes/default.rb", … } ] }
+populate_segments(Data, Metadata) ->
+    CB1 = lists:foldl(fun(Segment, CB) ->
+                              ej:set({Segment}, CB, [])
+                      end,
+                      Metadata,
+                      ?COOKBOOK_SEGMENTS),
+
+    lists:foldl(fun(File, CB) ->
+                        { Segment, Record } = remove_segment_from_filename(File),
+                        case lists:member(Segment, ?COOKBOOK_SEGMENTS) of
+                            true -> ej:set_p({Segment, new}, CB, Record);
+                            _ -> CB
+                        end
+                end,
+                CB1,
+                Data).
+
+ensure_v2_metadata(Ejson) ->
+    case ej:get({<<"all_files">>}, Ejson) of
+        undefined ->
+            Ejson1 = ej:set({<<"all_files">>}, Ejson, []),
+            lists:foldl(fun(Segment, CB) ->
+                                case ej:get({Segment}, CB) of
+                                    undefined -> CB;
+                                    Data ->
+                                        CB1 = populate_all_files(Segment, Data, CB),
```

Comment:
It looks like we replace all_files in populate_all_files above, so this fold would only keep the last segment.